shells/android/java/arcs/storage/service/StorageServiceConnection.kt

```diff
@@ -0,0 +1,144 @@
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.Job
+
+/** A [ConnectionFactory] is capable of creating a [StorageServiceConnection]. */
+typealias ConnectionFactory =
+    (StoreOptions<*, *, *>, ParcelableCrdtType) -> StorageServiceConnection
+
+/**
+ * Returns a default [ConnectionFactory] implementation which uses the provided [context] to bind to
+ * the [StorageService] and the provided [coroutineContext] as the parent for
+ * [StorageServiceConnection.connectAsync]'s [Deferred] return value.
+ */
+@Suppress("FunctionName")
+@ExperimentalCoroutinesApi
+fun DefaultConnectionFactory(
+    context: Context,
+    bindingDelegate: StorageServiceBindingDelegate = DefaultStorageServiceBindingDelegate(context),
+    coroutineContext: CoroutineContext = Dispatchers.Default
+): ConnectionFactory = { options, crdtType ->
+    StorageServiceConnection(bindingDelegate, options.toParcelable(crdtType), coroutineContext)
+}
+
+/** Defines an object capable of binding-to and unbinding-from the [StorageService]. */
+interface StorageServiceBindingDelegate {
+    fun bindStorageService(
+        conn: ServiceConnection,
+        flags: Int,
+        options: ParcelableStoreOptions
+    ): Boolean
+
+    fun unbindStorageService(conn: ServiceConnection)
+}
+
+/** Default implementation of the [StorageServiceBindingDelegate]. */
+class DefaultStorageServiceBindingDelegate(
+    private val context: Context
+) : StorageServiceBindingDelegate {
+    override fun bindStorageService(
+        conn: ServiceConnection,
+        flags: Int,
+        options: ParcelableStoreOptions
+    ): Boolean = context.bindService(StorageService.createBindIntent(context, options), conn, flags)
+
+    override fun unbindStorageService(conn: ServiceConnection) = context.unbindService(conn)
+}
+
+/** Object capable of managing a connection to the [StorageService]. */
+@UseExperimental(ExperimentalCoroutinesApi::class)
+class StorageServiceConnection(
+    /**
+     * Delegate which is responsible for actually initiating and tearing-down a binding to the
+     * [StorageService].
+     */
+    private val bindingDelegate: StorageServiceBindingDelegate,
+    /** Parcelable [StoreOptions] to pass to the [StorageService] when connecting. */
+    private val storeOptions: ParcelableStoreOptions,
+    /** Parent [CoroutineContext] for the [Deferred] returned by [connectAsync]. */
+    private val coroutineContext: CoroutineContext = Dispatchers.Default
+) : ServiceConnection {
+    private var needsDisconnect = false
+    private var service = atomic<CompletableDeferred<IStorageService>?>(null)
+
+    /** Whether or not the connection is active/alive. */
+    val isConnected: Boolean
+        get() = needsDisconnect &&
+            service.value?.let {
+                it.isCompleted && it.getCompleted().asBinder().isBinderAlive
+            } == true
```

Comment:
Probably don't need `== true` here?

Reply:
We do, since `?.let` is nullable, and `Boolean?` is not a valid input into a boolean expression... however, `Boolean? == Boolean` is a `Boolean`, and thus can go into the expression.

Reply:
Oh, right. Makes sense!